steel/src/history/mod.rs

```diff
@@ -0,0 +1,254 @@
+    /// Commitment for `state` to a Beacon Chain block root.
+    state_commit: GeneralizedBeaconCommit<STATE_ROOT_LEAF_INDEX>,
+}
+
+impl<H: EvmBlockHeader> BlockHeaderCommit<H> for HistoryCommit {
+    /// Generates a commitment that proves the given block header is included in the Beacon Chain's
+    /// history.
+    #[inline]
+    fn commit(self, header: &Sealed<H>, config_id: B256) -> Commitment {
+        // first, compute the beacon commit of the EVM execution
+        let initial_commitment = self.evm_commit.commit(header, config_id);
+        let (mut timestamp, version) = initial_commitment.decode_id();
+        // just a sanity check, a BeaconCommit will always have this version
+        assert_eq!(version, CommitmentVersion::Beacon as u16);
+
+        let mut beacon_root = initial_commitment.digest;
+        for state_commit in self.state_commits {
+            // verify that the previous commitment is valid wrt the current state
+            let state_root = state_commit.state.root();
+            let commitment_root =
+                BeaconRootsContract::get_from_state(state_commit.state, timestamp)
+                    .expect("Beacon roots contract failed");
+            assert_eq!(commitment_root, beacon_root, "Beacon root does not match");
+
+            // compute the beacon commitment of the current state
+            let (commit_ts, commit_beacon_root) = state_commit.state_commit.into_commit(state_root);
+            timestamp = U256::from(commit_ts);
+            beacon_root = commit_beacon_root;
+        }
+
+        Commitment::new(
+            CommitmentVersion::Beacon as u16,
+            timestamp.to(),
+            beacon_root,
+            initial_commitment.configID,
+        )
+    }
+}
+
+#[cfg(feature = "host")]
+mod host {
+    use super::*;
+    use crate::{
+        beacon::host::{client::BeaconClient, create_beacon_commit},
+        ethereum::EthBlockHeader,
+    };
+    use alloy::{network::Ethereum, providers::Provider, transports::Transport};
+    use alloy_primitives::{BlockNumber, Sealable};
+    use anyhow::{ensure, Context};
+    use url::Url;
+
+    impl HistoryCommit {
+        /// Creates a `HistoryCommit` from an EVM block header and a commitment header.
+        ///
+        /// This method fetches the necessary data from the Ethereum and Beacon chain to construct a `HistoryCommit`.
+        /// It iterates through blocks from the EVM header's number up to the commitment header's number, generating `StateCommit`s for each block in the range.
+        pub(crate) async fn from_headers<T, P>(
+            evm_header: &Sealed<EthBlockHeader>,
+            commitment_header: &Sealed<EthBlockHeader>,
+            rpc_provider: P,
+            beacon_url: Url,
+        ) -> anyhow::Result<Self>
+        where
+            T: Transport + Clone,
+            P: Provider<T, Ethereum>,
+        {
+            let client = BeaconClient::new(beacon_url.clone()).context("invalid URL")?;
```

Comment:
We might want to start with a check that `evm_header` is at least as old as `commitment_header`.

Reply:
Addressed in 836980a2203abd25e40933d7d1895563f21d1778